core/watcher/eventsource/namespace.go

```diff
@@ -26,44 +26,62 @@ type NamespaceQuery Query[[]string]
 type NamespaceWatcher struct {
 	*BaseWatcher
 
-	out chan []string
-
 	// TODO (manadart 2023-05-24): Consider making this plural (composite key)
 	// if/when it is supported by the change log table structure and stream.
-	namespace    string
 	initialQuery NamespaceQuery
-	changeMask   changestream.ChangeType
 
-	mapper Mapper
+	out        chan []string
+	filterOpts []changestream.SubscriptionOption
+	mapper     Mapper
 }
 
-// NewNamespaceWatcher returns a new watcher that receives changes from the
-// input base watcher's db/queue when changes in the namespace occur.
-// It emits the values from the namespace that changed.
+// NewNamespaceWatcher returns a new watcher that filters changes from the input
+// base watcher's db/queue. Change-log events will be emitted only if the filter
+// accepts them, and dispatching the notifications via the Changes channel. A
+// filter option is required, though additional filter options can be provided.
 func NewNamespaceWatcher(
-	base *BaseWatcher, namespace string, changeMask changestream.ChangeType, initialQuery NamespaceQuery,
-) watcher.StringsWatcher {
-	return NewNamespaceMapperWatcher(base, namespace, changeMask, initialQuery, defaultMapper)
+	base *BaseWatcher,
+	initialQuery NamespaceQuery,
+	filterOption FilterOption, filterOptions ...FilterOption,
+) (watcher.StringsWatcher, error) {
+	return NewNamespaceMapperWatcher(base, initialQuery, defaultMapper, filterOption, filterOptions...)
 }
 
-// NewNamespaceMapperWatcher returns a new watcher that receives changes
-// from the input base watcher's db/queue when changes in the namespace occur.
-// Values from the namespace that change are processed by the input mapper,
-// and based on the mapper's logic a subset of them (or none) may be emitted.
+// NewNamespaceMapperWatcher returns a new watcher that receives changes from
+// the input base watcher's db/queue. Change-log events will be emitted only if
+// the filter accepts them, and dispatching the notifications via the Changes
+// channel, once the mapper has processed them. Filtering of values is done
+// first by the filter, and then by the mapper. Based on the mapper's logic a
+// subset of them (or none) may be emitted. A filter option is required, though
+// additional filter options can be provided.
 func NewNamespaceMapperWatcher(
-	base *BaseWatcher, namespace string, changeMask changestream.ChangeType, initialQuery NamespaceQuery, mapper Mapper,
-) watcher.StringsWatcher {
+	base *BaseWatcher,
+	initialQuery NamespaceQuery, mapper Mapper,
+	filterOption FilterOption, filterOptions ...FilterOption,
+) (watcher.StringsWatcher, error) {
+	filters := append([]FilterOption{filterOption}, filterOptions...)
+	opts := make([]changestream.SubscriptionOption, len(filters))
+	for i, opt := range filters {
```

Comment:
Might as well check if the `opt` is nil too if you're doing other validation here.